### PR TITLE
[JUJU-2339] Remove macaroons for no cs on juju

### PR DESCRIPTION
**Pull request by cderici**
Remove the CharmStore macaroons as a side effort to retiring CharmStore from the client entirely. 

Relates to PRs such as #14968, #14985. 

https://warthogs.atlassian.net/browse/JUJU-2230 -- removing the CharmStore entirely from the client is coming up next.

## QA Steps

As CharmStore charms are not supported, this doesn't change any behavior.

[JUJU-2230]: https://warthogs.atlassian.net/browse/JUJU-2230?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apiserver/facades/client/charms/client.go [hunk 1]
```diff
@@ -236,24 +230,12 @@ func normalizeCharmOrigin(origin params.CharmOrigin, fallbackArch string) (param
...
 	})
 }
 
-// AddCharmWithAuthorization adds the given charm URL (which must include
-// revision) to the environment, if it does not exist yet. Local charms are
-// not supported, only charm store and charm hub URLs. See also AddLocalCharm().
-//
-// The authorization macaroon, args.CharmStoreMacaroon, may be
-// omitted, in which case this call is equivalent to AddCharm.
-func (a *API) AddCharmWithAuthorization(args params.AddCharmWithAuth) (params.CharmOriginResult, error) {
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/client.go:252]: You need to rev the facade to version 6, and keep this method for version 5. juju 3.0.x clients use the method. The 3.0.x client should still be able to deploy charmhub charms to a 3.1 controller, and this method is used.

File: apiserver/facades/controller/firewaller/firewaller.go [hunk 2]
```diff
@@ -307,26 +307,6 @@ func (f *FirewallerAPI) WatchIngressAddressesForRelations(relations params.Entit
 	return results, nil
 }
 
-// MacaroonForRelations returns the macaroon for the specified relations.
-func (f *FirewallerAPI) MacaroonForRelations(args params.Entities) (params.MacaroonResults, error) {
```
Comment (hmlanigan) [inline apiserver/facades/controller/firewaller/firewaller.go:311]: I think this is a different type of macaroon, not for charm store. It's related to CMR, which wouldn't be contacting the charm store. Should investigate this more.

File: state/charm.go [hunk 3]
```diff
@@ -756,39 +718,6 @@ func (c *Charm) IsPlaceholder() bool {
 	return c.doc.Placeholder
 }
 
-// Macaroon return the macaroon that can be used to request data about the charm
-// from the charmstore, or nil if the charm is not private.
-func (c *Charm) Macaroon() (macaroon.Slice, error) {
```
Comment (hmlanigan) [inline state/charm.go:761]: Add a charm to remove 
```
Macaroon     []byte `bson:"macaroon"`
```
from CharmDoc in state in 3.1

**Review by hmlanigan — CHANGES_REQUESTED** [review]

File: cmd/juju/application/refresher/refresher.go [hunk 4]
```diff
@@ -379,16 +378,10 @@ func (r *charmStoreRefresher) Refresh() (*CharmID, error) {
...
-		// for updating resources.
-		csMac, csErr := store.AuthorizeCharmStoreEntity(r.authorizer, newURL)
-		if csErr != nil && !strings.Contains(csErr.Error(), "404 NOT FOUND") {
-			return nil, errors.Trace(csErr)
-		}
+		// URL, origin for updating resources.
 		return &CharmID{
-			URL:      newURL,
-			Origin:   origin.CoreCharmOrigin(),
-			Macaroon: csMac,
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/refresher.go:391]: CharmID here is defined in refresher/interface.go, Macaroon can be removed from it now too

File: core/charm/repository/charmhub.go [hunk 5]
```diff
@@ -510,7 +498,7 @@ func (c *CharmHubRepository) getEssentialMetadataForBatch(reqs []corecharm.Metad
 	return metaRes, nil
 }
 
-func (c *CharmHubRepository) refreshOne(charmURL *charm.URL, origin corecharm.Origin, _ macaroon.Slice) (transport.RefreshResponse, error) {
+func (c *CharmHubRepository) refreshOne(charmURL *charm.URL, origin corecharm.Origin) (transport.RefreshResponse, error) {
 	// TODO(achilleasa): pass macaroons to client when charmhub rolls out support for private charms.
```
Comment (hmlanigan) [inline core/charm/repository/charmhub.go:502]: remove todo

File: core/charm/repository/charmhub.go [hunk 6]
```diff
@@ -389,26 +387,16 @@ func (c *CharmHubRepository) ListResources(charmURL *charm.URL, origin corecharm
 // a slice with the results. The results include the minimum set of metadata
 // that is required for deploying each charm.
 func (c *CharmHubRepository) GetEssentialMetadata(reqs ...corecharm.MetadataRequest) ([]corecharm.EssentialMetadata, error) {
+	// TODO (cderici): This grouping below might not be necessary anymore
+	// since we don't have any macaroons anymore
+
```
Comment (hmlanigan) [inline core/charm/repository/charmhub.go:392]: unrelated, couldn't select the lines for a pr comment - there is a TODO to be removed: lines 424-428

File: core/charm/repository/charmhub.go [hunk 7]
```diff
@@ -389,26 +387,16 @@ func (c *CharmHubRepository) ListResources(charmURL *charm.URL, origin corecharm
 // a slice with the results. The results include the minimum set of metadata
 // that is required for deploying each charm.
 func (c *CharmHubRepository) GetEssentialMetadata(reqs ...corecharm.MetadataRequest) ([]corecharm.EssentialMetadata, error) {
+	// TODO (cderici): This grouping below might not be necessary anymore
```
Comment (hmlanigan) [inline core/charm/repository/charmhub.go:390]: To answer the question, it looks like lines 390 to 419 can be deleted to me.

File: cmd/juju/application/store/store.go [hunk 8]
```diff
@@ -36,23 +32,17 @@ func AddCharmFromURL(client CharmAdder, curl *charm.URL, origin commoncharm.Orig
 // add the given charm URL to state. For non-public charm URLs, this function
 // also handles the macaroon authorization process using the given csClient.
```
Comment (hmlanigan) [inline cmd/juju/application/store/store.go:33]: header comment requires updates to remove macaroon

File: cmd/juju/application/deployer/charm.go [hunk 9]
```diff
@@ -186,7 +186,6 @@ func (d *deployCharm) deploy(
 			URL:    id.URL,
 			Origin: id.Origin,
 		},
-		d.csMac,
```
Comment (hmlanigan) [inline cmd/juju/application/deployer/charm.go:189]: csMac can also be removed from the deployCharm struct now, deployer/charm.go ln 42

File: cmd/juju/application/deployer/bundlehandler.go [hunk 10]
```diff
@@ -240,7 +240,6 @@ func makeBundleHandler(defaultCharmSchema charm.Schema, bundleData *charm.Bundle
 		data:                 bundleData,
 		bundleURL:            spec.bundleURL,
 		unitStatus:           make(map[string]string),
-		macaroons:            make(map[charm.URL]*macaroon.Macaroon),
```
Comment (hmlanigan) [inline cmd/juju/application/deployer/bundlehandler.go:243]: remove from struct definition as well.

File: cmd/juju/application/store/interface.go [hunk 11]
```diff
@@ -38,5 +36,5 @@ type MacaroonGetter interface {
 // CharmsAPI is functionality needed by the CharmAdapter from the Charms API.
```
Comment (hmlanigan) [inline cmd/juju/application/store/interface.go:36]: The MacaroonGetter interface above be removed.

File: cmd/juju/application/deploy_test.go [hunk 12]
```diff
@@ -173,7 +173,6 @@ func (s *DeploySuiteBase) SetUpTest(c *gc.C) {
 	s.AddCleanup(func(*gc.C) { s.CmdBlockHelper.Close() })
 	s.DeployResources = func(applicationID string,
 		chID resources.CharmID,
```
Comment (hmlanigan) [inline cmd/juju/application/deploy_test.go:175]: deploy_test still imports macaroon.v2, so some references still exist.

File: apiserver/facades/client/charms/client.go (hunk 1 above)
Comment (hmlanigan) [reply apiserver/facades/client/charms/client.go:252]: No problem deploying a ch charm from a 3.0 client to a controller running this code.

**Review by hmlanigan — COMMENTED** [review]
I think we can stop using a *httpbakery.Client, however that might be a piece for when we stop using the charmstore client in general.

File: cmd/juju/application/store/store.go [hunk 13]
```diff
@@ -36,23 +32,17 @@ func AddCharmFromURL(client CharmAdder, curl *charm.URL, origin commoncharm.Orig
...
-			return nil, nil, commoncharm.Origin{}, errors.Trace(err)
-		}
-		m, err := AuthorizeCharmStoreEntity(cs, curl)
-		if err != nil {
-			return nil, nil, commoncharm.Origin{}, common.MaybeTermsAgreementError(err)
+			return nil, commoncharm.Origin{}, errors.Trace(err)
 		}
-		if resultOrigin, err = client.AddCharmWithAuthorization(curl, origin, m, force); err != nil {
-			return nil, nil, commoncharm.Origin{}, errors.Trace(err)
+		if resultOrigin, err = client.AddCharm(curl, origin, force); err != nil {
```
[the remark this answers is not in the document]
Comment (cderici) [reply cmd/juju/application/store/store.go:41]: Yeah that's basically [eta-equivalent ](https://en.wiktionary.org/wiki/eta_conversion#English) to `AddCharm`, we can just use that instead :+1:

File: core/charm/repository/charmhub.go [hunk 14]
```diff
@@ -389,40 +387,19 @@ func (c *CharmHubRepository) ListResources(charmURL *charm.URL, origin corecharm
 // a slice with the results. The results include the minimum set of metadata
 // that is required for deploying each charm.
 func (c *CharmHubRepository) GetEssentialMetadata(reqs ...corecharm.MetadataRequest) ([]corecharm.EssentialMetadata, error) {
```
Comment (hmlanigan) [inline core/charm/repository/charmhub.go:389]: I believe you can delete lines 392 to and including 431. GetEssentialMetadata does batching which is no longer required as there are no macaroons to batch. getEssentialMetadataForBatch does the real work still required.

Then the TODO at 436 can be removed.

File: core/charm/repository/charmhub.go (hunk 5 above)
Comment (hmlanigan) [reply core/charm/repository/charmhub.go:491]: todo still needs to be removed.

File: core/charm/downloader/downloader.go [hunk 15]
```diff
@@ -206,7 +206,6 @@ func (d *Downloader) downloadAndHash(charmURL *charm.URL, requestedOrigin corech
 		Size:         size,
 		LXDProfile:   chArchive.LXDProfile(),
 		SHA256:       sha,
-		Macaroons:    macaroons,
```
Comment (hmlanigan) [inline core/charm/downloader/downloader.go:209]: remove Macaroons from the DownloadedCharm type definition

**Review by hmlanigan — APPROVED** [review]
LGTM. Removing functionality can be hard work!

**Pull request merged**